lib/event_sourcery/event_store/event_builder.rb

```diff
@@ -8,12 +8,16 @@ def initialize(event_type_serializer:)
       def build(event_data)
         type = event_data.fetch(:type)
         klass = event_type_serializer.deserialize(type)
-        klass.new(event_data)
+        upcast(klass.new(event_data))
       end
 
       private
 
       attr_reader :event_type_serializer
+
+      def upcast(event)
+        event.class.upcast(event)
+      end
```

Comment:
I see upcasting as a transformation on an event body (not including any other event attributes). Is there a need to upcast anything but the event body?

Reply:
More context here: https://github.com/envato/event_sourcery/pull/190#issuecomment-355466774

I don't have a use case for it yet but I can imagine actually changing the event class could make the implementation clearer. I think this API just provides more flexibility.

Reply:
Yeah I guess type would be a useful thing to be able to change. `ItemCreated` -> `ItemCreatedV2`.

Reply:
@stevehodgkiss I wonder if we could have used this in a scenario like when we migrated Identity's old billing_details_overridden events to new versions (rather than having to emit new version events for all existing ones)? 🤔

Reply:
Agreed. This provides a mechanism for updating versions of historic events where the change is simple enough that you don't need to reproject the event stream.